com.unity.ml-agents/Runtime/SideChannels/SideChannel.cs

```diff
@@ -28,16 +31,212 @@ public abstract class SideChannel
         /// Is called by the communicator every time a message is received from Python by the SideChannel.
         /// Can be called multiple times per simulation step if multiple messages were sent.
         /// </summary>
-        /// <param name="data"> the payload of the message.</param>
-        public abstract void OnMessageReceived(byte[] data);
+        /// <param name="msg">The incoming message.</param>
+        public abstract void OnMessageReceived(IncomingMessage msg);
 
         /// <summary>
         /// Queues a message to be sent to Python during the next simulation step.
         /// </summary>
         /// <param name="data"> The byte array of data to be sent to Python.</param>
-        protected void QueueMessageToSend(byte[] data)
+        protected void QueueMessageToSend(OutgoingMessage msg)
+        {
+            MessageQueue.Add(msg.ToByteArray());
+        }
+    }
+
+    /// <summary>
```

Comment:
Put in a separate file

Reply:
Yeah, makes sense, wasn't sure where to draw the line.